logger/src/dal.rs

```diff
@@ -0,0 +1,292 @@
+            LogLevel::Info => Self::Info,
+            LogLevel::Warn => Self::Warn,
+            LogLevel::Error => Self::Error,
+        }
+    }
+}
+
+impl From<SeverityNumber> for LogLevel {
+    fn from(severity: SeverityNumber) -> Self {
+        match severity.into() {
+            log::Level::Trace => Self::Trace,
+            log::Level::Debug => Self::Debug,
+            log::Level::Info => Self::Info,
+            log::Level::Warn => Self::Warn,
+            log::Level::Error => Self::Error,
+        }
+    }
+}
+
+impl Log {
+    /// Try to get a log from an OTLP [ResourceSpans]
+    pub fn try_from_scope_span(resource_spans: ResourceSpans) -> Option<Vec<Self>> {
+        let ResourceSpans {
+            resource,
+            scope_spans,
+            schema_url: _,
+        } = resource_spans;
+
+        // TODO: we should get both of these attributes in the same function and avoid this clone.
+        let resource = resource?;
+        let shuttle_service_name = get_attribute(resource.clone().attributes, "service.name")?;
+
+        // Try to get the deployment_id from the resource attributes, this will be the case for the runtimes,
+        // they add the deployment_id to the otlp tracer config.
+        // TODO: should this be named "deployment.id" to conform to otlp standard?
+        let deployment_id = get_attribute(resource.attributes, "deployment_id");
+
+        let logs = scope_spans
+            .into_iter()
+            .flat_map(|scope_spans| {
+                let ScopeSpans {
+                    spans,
+                    schema_url: _,
+                    ..
+                } = scope_spans;
+
+                let events: Vec<_> = spans
+                    .into_iter()
+                    .flat_map(|span| {
+                        Self::try_from_span(span, &shuttle_service_name, deployment_id.clone())
+                    })
+                    .flatten()
+                    .collect();
+
+                Some(events)
+            })
+            .flatten()
+            .collect();
+
+        Some(logs)
+    }
+
+    /// Try to get self from an OTLP [Span]. Also enrich it with the shuttle service name and deployment id.
+    fn try_from_span(
+        span: Span,
+        shuttle_service_name: &str,
+        deployment_id: Option<String>,
+    ) -> Option<Vec<Self>> {
+        // If we didn't find the id in the resource span, check the inner spans.
+        let deployment_id = deployment_id.or(get_attribute(span.attributes, "deployment_id"))?;
+
+        let logs = span
+            .events
+            .into_iter()
+            .flat_map(|event| {
+                let message = event.name;
+
+                let mut fields = from_any_value_kv_to_serde_json_map(event.attributes);
+                fields.insert("message".to_string(), message.into());
+
+                // Every event should have a level field, but we store it in the level column,
+                // so we can remove it from the event fields.
```

Comment:
Mind explaining this more please :smile:

Reply:
Clarified the comment now.